src/types/transaction.zig

```diff
@@ -1,139 +1,211 @@
+
+    try self.serializeToSlice(ret);
+
+    return ret;
+}
+
+pub fn deserializeReader(allocator: std.mem.Allocator, r: anytype) !Self {
+    comptime {
+        if (!std.meta.hasFn(@TypeOf(r), "readInt")) @compileError("Expects r to have fn 'readInt'.");
+        if (!std.meta.hasFn(@TypeOf(r), "readNoEof")) @compileError("Expects r to have fn 'readNoEof'.");
+        if (!std.meta.hasFn(@TypeOf(r), "readAll")) @compileError("Expects r to have fn 'readAll'.");
+        if (!std.meta.hasFn(@TypeOf(r), "readByte")) @compileError("Expects r to have fn 'readByte'.");
     }
 
-    /// Add an output to the transaction
-    pub fn addOutput(self: *Transaction, value: i64, script_pubkey: Script) !void {
-        var new_script = try Script.init(self.allocator);
-        try new_script.push(script_pubkey.bytes);
-        try self.outputs.append(Output{
-            .value = value,
-            .script_pubkey = new_script,
-        });
+    var tx: Self = try Self.init(allocator);
+
+    tx.version = try r.readInt(i32, .little);
+
+    const compact_input_len = try CompactSizeUint.decodeReader(r);
+    const inputs = try allocator.alloc(Input, compact_input_len.value());
+    errdefer allocator.free(inputs);
+
+    var i: usize = 0;
+    while (i < compact_input_len.value()) : (i += 1) {
+        const input = try Input.deserializeReader(allocator, r);
+
+        inputs[i] = input;
     }
+    tx.inputs = inputs;
+
+    const compact_output_len = try CompactSizeUint.decodeReader(r);
+    const outputs = try allocator.alloc(Output, compact_output_len.value());
+    errdefer allocator.free(outputs);
 
-    /// Calculate the transaction hash
-    pub fn hash(self: *const Transaction) Hash {
-        var h: [32]u8 = undefined;
-        std.crypto.hash.sha2.Sha256.hash(@as([]const u8, std.mem.asBytes(&self.version)), &h, .{});
-        return Hash{ .bytes = h };
+    var j: usize = 0;
+    while (j < compact_output_len.value()) : (j += 1) {
+        const output = try Output.deserializeReader(allocator, r);
+
+        outputs[j] = output;
     }
+    tx.outputs = outputs;
+    tx.lock_time = try r.readInt(u32, .little);
+
+    return tx;
+}
 
-    /// Calculate the virtual size of the transaction
-    pub fn virtual_size(self: *const Transaction) usize {
-        // This is a simplified size calculation. In a real implementation,
-        // you would need to account for segregated witness data if present.
-        var size: usize = 8; // Version (4 bytes) + LockTime (4 bytes)
-        size += self.inputs.items.len * 41; // Simplified input size
-        size += self.outputs.items.len * 33; // Simplified output size
-        return size;
+/// Deserialize bytes into Self
+pub fn deserializeSlice(allocator: std.mem.Allocator, bytes: []const u8) !Self {
+    var fbs = std.io.fixedBufferStream(bytes);
+    return try Self.deserializeReader(allocator, fbs.reader());
+}
+
+/// Initialize a new transaction
+pub fn init(allocator: std.mem.Allocator) !Self {
```

Comment:
Is this really used?

Reply:
It it's only for tests don't make it public

Reply:
It's used in mempool's test, and in storage.zig l40